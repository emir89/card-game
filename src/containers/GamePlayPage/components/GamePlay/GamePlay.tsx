import { useState, useEffect } from 'react';
import style from './gamePLay.module.css';
import { getWinningCardImage } from 'util/gameLogic';
import NotificationBar from 'commonComponents/NotificationBar/NotificationBar';
import SelectedCards from '../SelectedCards/SelectedCards';
import DeckOfCards from '../DeckOfCards/DeckOfCards';
import { hashString } from 'util/hash';
import GameWinner from '../GameWinner/GameWinner';
import { getCardValue } from 'util/card';
import { Card, SelectedCard, WinnerByRound } from '../../types';

type GamePlayProps = {
  drawCards: Card[] | null;
  numOfPlayers: number;
};

const GamePlay = ({ drawCards, numOfPlayers }: GamePlayProps) => {
  const [selectedCards, setSelectedCards] = useState<SelectedCard[]>([]);
  const [winnersByRound, setWinnersByRound] = useState<WinnerByRound[]>([]);
  const [numOfRounds, setNumOfRounds] = useState<number>(0);
  const [playersDecks, setPlayersDecks] = useState<Card[][]>([]);
  const [gameWinners, setGameWinners] = useState<WinnerByRound[]>([]);

  const showGameStartNotification = numOfRounds === 0 && selectedCards?.length === 0;

  // Helper: Add selected card
  const addSelectedCards = (player: number, selectedCard: Card) => {
    setSelectedCards((prev) => [
      ...prev,
      {
        image: selectedCard.image,
        value: getCardValue(selectedCard),
        player,
      },
    ]);
  };

  // Helper: Remove selected cards
  const removeSelectedCards = () => {
    setSelectedCards([]);
  };

  // Helper: Update players decks
  const handleUpdatePlayersDecks = (
    player?: number,
    selectedCard?: Card,
    playerDecksOverride?: Card[][]
  ) => {
    if (!player) {
      setPlayersDecks(playerDecksOverride ?? []);
      return;
    }
    if (selectedCard) {
      setPlayersDecks((prev) =>
        prev.map((deck, idx) =>
          idx === player - 1 ? deck.filter((card) => card.image !== selectedCard.image) : deck
        )
      );
    }
  };

  // Helper: Handle round winner
  const handleRoundWinner = () => {
    if (selectedCards.length === 0) return;
    const sortedSelectedCards = selectedCards.slice().sort((b, a) => {
      if (a.value === b.value) {
        return a.player > b.player ? 1 : -1;
      }
      return a.value - b.value;
    });
    const winner = sortedSelectedCards[0];
    setWinnersByRound((prev) => {
      const existing = prev.find((w) => w.player === winner.player);
      if (existing) {
        return prev.map((w) =>
          w.player === winner.player
            ? { ...w, score: w.score + winner.value, byRound: numOfRounds + 1 }
            : w
        );
      } else {
        return [
          ...prev,
          {
            player: winner.player,
            score: winner.value,
            byRound: numOfRounds + 1,
          },
        ];
      }
    });
  };

  // Helper: Handle number of rounds
  const handleNumberOfRounds = () => {
    setNumOfRounds((prev) => prev + 1);
  };

  // useEffect to determine overall winner after last round
  useEffect(() => {
    if (numOfRounds === 10) {
      // Sort the winners by score in descending order
      const sortedWinnersByRound = winnersByRound.slice().sort((b, a) => {
        return a.score > b.score ? 1 : -1;
      });
      // Gather all players who have the highest score (handle ties)
      const gameWinners = sortedWinnersByRound.reduce<WinnerByRound[]>((acc, curr, i) => {
        if (i === 0) {
          acc.push(curr);
        } else {
          if (acc[0].score === curr.score) {
            acc.push(curr);
          }
        }
        return acc;
      }, []);
      setGameWinners(gameWinners);
    }
  }, [winnersByRound, numOfRounds]);

  /**
   * Handles click on a card
   *
   * @param player
   * @param selectedCard
   */
  const onClickCard = (player: number, selectedCard: Card) => {
    handleUpdatePlayersDecks(player, selectedCard);
    addSelectedCards(player, selectedCard);
  };

  // Handles the logic for progressing to the next player's turn or moving to the next round.
  const handleNextPlayerOrRound = (selectedCards: SelectedCard[], playersDecks: Card[][]) => {
    // Count how many cards have been selected in the current round
    const numOfSelectedCards = selectedCards.length;

    // If not all players have selected a card yet, move to the next player's turn
    if (numOfSelectedCards < numOfPlayers) {
      setTimeout(() => {
        // Determine the next player's index (1-based)
        const player = numOfSelectedCards + 1;
        // Get the current player's deck
        const playerDeck = playersDecks[player - 1];
        if (!playerDeck) return; // If no deck found, exit

        // Map the player's deck to the card objects with image, value, and player number
        const playerCards = playerDeck.map((deck) => ({
          image: deck.image,
          value: String(getCardValue(deck)),
          player,
        }));
        // Randomly select a card for the player
        const selectedCard = playerCards[Math.floor(Math.random() * playerCards.length)];
        // Trigger the card selection for the current player
        onClickCard(player, selectedCard);
      }, 1000); // Wait 1 second before next player's turn (for UX pacing)
    } else if (numOfSelectedCards === numOfPlayers) {
      // All players have selected a card, determine the round winner
      handleRoundWinner();
      // Update the number of rounds played
      handleNumberOfRounds();

      // After a short delay, remove selected cards to prepare for the next round
      setTimeout(() => {
        removeSelectedCards();
      }, 1000);
    }
  };

  // Handles next player turn or round logic (triggered by state changes)
  useEffect(() => {
    if (selectedCards.length > 0) {
      handleNextPlayerOrRound(selectedCards, playersDecks);
    }
  }, [selectedCards, numOfPlayers, playersDecks]);

  // Initialize players decks once when drawCards is available and playersDecks is empty
  if (drawCards && playersDecks.length === 0) {
    const chunkSize = 10;
    const playersCards: Card[][] = [];
    for (let i = 0; i < drawCards.length; i += chunkSize) {
      playersCards.push(drawCards.slice(i, i + chunkSize));
    }
    handleUpdatePlayersDecks(undefined, undefined, playersCards);
  }

  // Use the shared helper for determining the winning card image
  const winningCardImage = getWinningCardImage(selectedCards, numOfPlayers);

  return (
    <div>
      {showGameStartNotification && (
        <NotificationBar>
          <p className={style.NotificationBarMessage}>
            Welcome. To start the game, click on one of the cards from your deck. Good Luck!
          </p>
        </NotificationBar>
      )}
      <SelectedCards
        selectedCards={selectedCards}
        winningCardImage={winningCardImage}
      />
      <div className={style.CircleTable}>
        {playersDecks?.map((deck, i) => {
          return (
            <DeckOfCards
              data-testid="deckOfCards"
              cards={deck}
              players={numOfPlayers}
              handleClickCard={onClickCard}
              winnersByRound={winnersByRound}
              numOfSelectedCards={selectedCards.length}
              playerOrder={i + 1}
              currentPlayer={selectedCards.length + 1}
              key={hashString(`deck-player-${i + 1}-${deck[0]?.image || ''}`)}
            />
          );
        })}
      </div>
      {gameWinners.length > 0 ? <GameWinner gameWinners={gameWinners} /> : null}
    </div>
  );
};

export default GamePlay;
