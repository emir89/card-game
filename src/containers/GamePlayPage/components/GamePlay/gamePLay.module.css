.DeckOfCardsWrapper,
.GameWrapper {
    display: flex;
}

.NotificationBarMessage {
    font-weight: bold;
}
