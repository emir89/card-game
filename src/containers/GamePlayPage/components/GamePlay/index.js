import React, {useState, useEffect} from 'react';
import style from "./gamePLay.module.css";
import NotificationBar from "../../../../commonComponents/NotificationBar";
import SelectedCards from "../SelectedCards";
import DeckOfCards from "../DeckOfCards";
import GameWinner from "../GameWinner";
import {
    handleRoundWinner,
    handleNumberOfRounds,
    handleAddSelectedCards,
    handleRemoveSelectedCards,
    handleUpdatePlayersDecks
} from "../../gamePlayActions";
import {getCardValue} from "../../../../util/card";
import PropTypes from 'prop-types';


function GamePlay({
  drawCards,
  numOfPlayers,
  dispatch,
  winnersByRound,
  numOfRounds,
  playersDecks,
  selectedCards,
  history
}) {
    let [gameWinners, setGameWinners] = useState([]);
    const numOfWinnersByRound = winnersByRound?.length;

    /**
     * Handles click on a card
     *
     * @param player
     * @param selectedCard
     */
    function onClickCard(player, selectedCard) {
        dispatch(handleUpdatePlayersDecks(player, selectedCard));
        dispatch(handleAddSelectedCards(selectedCard, player));
    }

    /**
     * Sets round winner info
     */
    function setRoundWinnerInfo() {
        if (numOfWinnersByRound === 0)  return null;
        const roundWinner = winnersByRound[numOfWinnersByRound - 1];

        return (
            <p className={style.NotificationBarMessage}>
                {`Round winner: Player ${roundWinner.player} 
                        with score of ${roundWinner.byRound}`}
            </p>
        );
    }

    // Removes selected cards
    useEffect(() => {
        let timer;
        if (selectedCards?.length === numOfPlayers) {
            timer = setTimeout(() => dispatch(handleRemoveSelectedCards(), 1000));
        }

        return () => {
            clearTimeout(timer);
        }
    }, [selectedCards, numOfPlayers, dispatch]);

    // Sets players decks
    useEffect(() => {
        if (drawCards) {
            const clonedCards = drawCards.cards.slice(0);

            let chunkSize = 10;
            let playersCards = clonedCards.map((card, i) => {
                return i % chunkSize === 0 ? clonedCards.slice(i, i + chunkSize) : null;
            }).filter(card => card);
            dispatch(handleUpdatePlayersDecks(null, null, playersCards));
        }
    }, [
        drawCards,
        dispatch
    ]);

    // handles winner by round
    useEffect(() => {
        if (selectedCards?.length) {
            let timer;
            const numOfSelectedCards = selectedCards.length;
            if (numOfSelectedCards < numOfPlayers) {
                timer = setTimeout(
                    () => {
                        const player = numOfSelectedCards + 1;
                        const playerCards = playersDecks[player - 1].map(deck => {
                            return {
                                image: deck.image,
                                value: getCardValue(deck),
                                player
                            }
                        });
                        const selectedCard = playerCards[Math.floor(Math.random() * playerCards.length)];
                        onClickCard(player, selectedCard);
                    },
                    1000);
            } else if (numOfSelectedCards === numOfPlayers) {
                dispatch(handleRoundWinner());
                dispatch(handleNumberOfRounds());
            }

            return () => {
                clearTimeout(timer);
            }
        }
    }, [
        selectedCards,
        numOfPlayers,
        playersDecks,
        dispatch
    ]);

    // handles game winner
    useEffect(() => {
        if (numOfRounds === 10) {
            const sortedWinnersByRound = winnersByRound.slice().sort((b, a) => {
                return a.score > b.score ? 1 : -1;
            });

            const gameWinners = sortedWinnersByRound.reduce((acc, curr, i) => {
                if (i === 0) {
                    acc.push(curr);
                } else {
                    if (acc[0].score === curr.score) {
                        acc.push(curr)
                    }
                }

                return acc;

            }, []);

            setGameWinners(gameWinners);
        }
    }, [
        numOfRounds,
        winnersByRound
    ]);

    return (
        <div className={style.GameWrapper}>
            <NotificationBar
                isVisible={numOfRounds === 0 && selectedCards?.length === 0}
            >
                <p className={style.NotificationBarMessage}>
                    Welcome. To start the game, click on one of the cards from your deck. Good Luck!
                </p>
            </NotificationBar>
            {winnersByRound &&
            <NotificationBar
                data-testid="winnersByRound"
                isVisible={selectedCards?.length === numOfPlayers}
                timeout={2000}
            >{
                setRoundWinnerInfo()
            }
            </NotificationBar>}
            <SelectedCards selectedCards={selectedCards}/>
            <div className={style.DeckOfCardsWrapper}>
                {
                    playersDecks?.map((deck, i) => {
                        return (
                            <DeckOfCards
                                data-testid="deckOfCards"
                                cards={deck}
                                players={numOfPlayers}
                                handleClickCard={onClickCard}
                                winnersByRound={winnersByRound}
                                numOfSelectedCards={selectedCards.length}
                                playerOrder={i + 1}
                                currentPlayer={selectedCards.length + 1}
                                key={i}
                            />
                        );
                    })
                }
            </div>
            {gameWinners.length > 0 &&
            <GameWinner
                gameWinners={gameWinners}
                history={history}
            />
            }
        </div>
    );
}

GamePlay.propTypes = {
    drawCards: PropTypes.object.isRequired,
    numOfPlayers: PropTypes.number.isRequired,
    history: PropTypes.object
};

export default GamePlay;
