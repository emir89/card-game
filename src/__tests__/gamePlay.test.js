import React from "react";
import { render, fireEvent, waitFor, cleanup } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect'
import GamePlay from '../containers/GamePlayPage/components/GamePlay';
import renderer from 'react-test-renderer';

const dummyData = {
    cards: [
        {
            image: 'foo1.png'
        },
        {
            image: 'foo2.png'
        },
        {
            image: 'foo3.png'
        },
        {
            image: 'foo4.png'
        },
        {
            image: 'foo5.png'
        },
        {
            image: 'foo6.png'
        },
        {
            image: 'foo7.png'
        },
        {
            image: 'foo8.png'
        },
        {
            image: 'foo9.png'
        },
        {
            image: 'foo10.png'
        },
        {
            image: 'foo11.png'
        },
        {
            image: 'foo12.png'
        },
        {
            image: 'foo13.png'
        },
        {
            image: 'foo14.png'
        },
        {
            image: 'foo15.png'
        },
        {
            image: 'foo16.png'
        },
        {
            image: 'foo17.png'
        },
        {
            image: 'foo18.png'
        },
        {
            image: 'foo19.png'
        },
        {
            image: 'foo20.png'
        }
    ]
};

describe('Game Play page', () => {
    afterEach(cleanup);

    test('It loads decks of cards', () => {
        const {getByText} = render(<GamePlay drawnCards={dummyData} numOfPlayers={2} />);

        waitFor(() => {
            expect(getByText('You')).toBeTruthy();
        });
    });

    test('It chooses a card from the deck', () => {
        const {getAllByTestId, getByTestId} = render(<GamePlay drawnCards={dummyData} numOfPlayers={2}/>);
        waitFor(() => {
            expect(getByTestId('deckOfCards')).toBeInTheDocument();
            fireEvent.click(getAllByTestId('clickableCards')[0]);

            expect(getByTestId('selectedCard')).toBeInTheDocument();
            expect(getAllByTestId('clickableCards')).toHaveLength(9);
        });
    });

    /////// SNAPSHOT ////////
    test('It matches snapshot', () => {
        const tree = renderer.create(<GamePlay drawnCards={dummyData} numOfPlayers={2} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
