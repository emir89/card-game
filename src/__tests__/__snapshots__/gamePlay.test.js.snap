// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Play page It matches snapshot 1`] = `
<div
  className="GameWrapper"
>
  <div
    className="CenterWrapper"
  />
  <div
    className="DeckOfCardsWrapper"
  />
</div>
`;
